crates/sqlbuiltins/src/functions/scalars/openai.rs

```diff
@@ -0,0 +1,252 @@
+use tokio::runtime::Handle;
+use tokio::task;
+
+use crate::functions::{BuiltinScalarUDF, ConstBuiltinFunction};
+const DEFAULT_CREDENTIAL_LOCATION: Lazy<&[&str]> =
+    Lazy::new(|| &["@creds", "openai", "openai_default_credential", "api_key"]);
+// This is a placeholder for empty values in the input array
+// The openai API does not accept empty strings, so we use this to represent NULL/"" values
+const EMPTY_PLACEHOLDER: &str = "NULL";
+
+pub struct OpenAIEmbed;
+
+impl ConstBuiltinFunction for OpenAIEmbed {
+    const NAME: &'static str = "openai_embed";
+    const DESCRIPTION: &'static str = "Embeds text using OpenAI's API. 
+    Available models: 'text-embedding-3-small', 'text-embedding-ada-002', 'text-embedding-3-large' default: 'text-embedding-3-small'
+    Note: This function requires an API key. You can pass it as the first argument or set it as a stored credential.
+    If no API key is provided, the function will attempt to use the stored credential 'openai_default_credential'";
+    const EXAMPLE: &'static str =
+        "openai_embed(@creds.openai.my_openai.api_key, 'text-embedding-3-small', 'Hello, world!')";
+    const FUNCTION_TYPE: FunctionType = FunctionType::Scalar;
+    fn signature(&self) -> Option<Signature> {
+        Some(Signature::new(
+            // args: <API_KEY>, <MODEL>, <TEXT>
+            TypeSignature::OneOf(vec![
+                // openai_embed('<text>')  -- uses default model and credential
+                TypeSignature::Exact(vec![DataType::Utf8]),
+                // openai_embed('<model>', '<text>') -- uses default credential
+                TypeSignature::Exact(vec![DataType::Utf8, DataType::Utf8]),
+                // openai_embed('api_key', '<model>', '<text>')  --uses provided api key and model
+                TypeSignature::Exact(vec![DataType::Utf8, DataType::Utf8, DataType::Utf8]),
+            ]),
+            // This is a volatile function because it makes an external API call.
+            // Additionally, the openai API key may change or expire at any time.
+            Volatility::Volatile,
+        ))
+    }
+}
+#[derive(Debug)]
+enum EmbeddingModel {
+    TextEmbedding3Small,
+    TextEmbeddingAda002,
+    TextEmbedding3Large,
+}
+impl EmbeddingModel {
+    fn len(&self) -> i32 {
+        match self {
+            EmbeddingModel::TextEmbedding3Small => 1536,
+            EmbeddingModel::TextEmbeddingAda002 => 1536,
+            EmbeddingModel::TextEmbedding3Large => 3072,
+        }
+    }
+}
+impl FromStr for EmbeddingModel {
+    type Err = datafusion::error::DataFusionError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "text-embedding-3-small" => Ok(EmbeddingModel::TextEmbedding3Small),
+            "text-embedding-ada-002" => Ok(EmbeddingModel::TextEmbeddingAda002),
+            "text-embedding-3-large" => Ok(EmbeddingModel::TextEmbedding3Large),
+            _ => Err(DataFusionError::Execution("Invalid model name".to_string())),
```

Comment:
Would be nice if we could return the available models in this error to keep the user from needing to guess.